proposals/p3980.md

```diff
@@ -0,0 +1,696 @@
+defined by it being in the same lexical scope), keywords should be duplicated to
+the definition. This is not proposed as a rule because it's not clear whether
+we'll generally follow it, but it's why this particular choice is taken.
+
+### Separate require-direct-import from non-owning declarations
+
+At present, an `extern` modifier on an owning declaration serves two purposes:
+
+1.  Indicates that a non-owning `extern library` declaration _can_ exist.
+2.  Indicates the declaration must be directly imported in order to be complete.
+
+This means that:
+
+-   The presence of `extern` on an owning declaration cannot be used to
+    determine whether a non-owning declaration exists.
+    -   Because the location of a non-owning declaration isn't explicit in the
+        owning code, this may lead to a developer failing to find the non-owning
+        declaration and misunderstanding that as the non-existence of a
+        non-owning declaration.
+-   Libraries which happen to be imported by the owning declaration may freely
+    add or remove non-owning `extern library` declarations without modifying the
+    owning library.
+
+We could give distinct syntax to the two purposes, so that they could be managed
+separately. The preference at present is to use a single syntax for both
+purposes, rather than emphasizing control or correspondence.
+
+### Other `extern` syntaxes
+
+[Issue #3986](https://github.com/carbon-language/carbon-lang/issues/3986)
+discussed other syntaxes for `extern` + `extern library`. These were mainly
+`has_extern`/`is_extern`/`externed` + `extern`.
+
+Breaking down `extern`, there are two features which could have been provided
+separately:
+
+1.  Declaring an entity has a forward declaration in a separate library.
+    -   Also, declaring that forward declaration in a separate library:
+        `extern library "<owning_library>"`.
+2.  Declaring an entity must be imported directly.
+
+Although (1) must depend on (2), a different design could provide (2) without
+making (1) possible, for example with different keywords to differentiate
+between intended usage (`has_extern class C;` meaning (1) and (2), `must_import`
+meaning (2) only). However, the `extern` keyword approach means developers have
+all or nothing.
+
+Considering that, the trade-offs are viewed as:
+
+-   The primary motivation is to provide feature (1).
+-   Leads consider it acceptable to support (2) separate from (1), and don't see
+    a problem with making (1) optional to achieve this (in other words, making
+    the `extern library "<owning_library>"` declaration optional).
+    -   It's okay that that `extern library "<owning_library>"` can be added and
+        removed from imported libraries without modifying the owning library.
+    -   If a developer considers it important to disambiguate the intended use
+        of a declaration `extern class C;` and whether there should be a
+        declaration in a separate library, they can add comments.
```

Comment:
Some small tweaks and a bit more motivation:

```suggestion
-   Leads wanted a syntax on the owning declaration that states something
    positive about the owning declaration itself, rather than expressing that
    other declarations exist, which suggests that the syntax on the owning
    declaration should provide feature (2).
-   Leads consider it valuable, though secondary, to support (2) separate from (1), and
    find it acceptable to make (1) optional to achieve this (in other words, making
    the `extern library "<owning_library>"` declaration optional).
    -   It's okay that that `extern library "<owning_library>"` can be added and
        removed from imported libraries without modifying the owning library.
    -   If a developer considers it important to disambiguate the intended use
        of a declaration `extern class C;` and whether there should be a
        declaration in a separate library, they can add comments.
```

... then I think this reasonably captures our discussions.